src/sources/amqp.rs

```diff
@@ -231,9 +232,115 @@ fn populate_event(
     log_namespace.insert_source_metadata("amqp", log, keys.offset_key, "offset", keys.delivery_tag);
 }
 
-/// Runs the AMQP source involving the main loop pulling data from the server.
+/// Receives an event from `AMQP` and pushes it along the pipeline.
+async fn receive_event(
+    config: &AmqpSourceConfig,
+    out: &mut SourceSender,
+    log_namespace: LogNamespace,
+    finalizer: Option<&UnorderedFinalizer<FinalizerEntry>>,
+    msg: Delivery,
+) -> Result<(), ()> {
+    let payload = Cursor::new(Bytes::copy_from_slice(&msg.data));
+    let mut stream = FramedRead::new(payload, config.decoder(log_namespace));
+
+    // Extract timestamp from amqp message
+    let timestamp = msg
+        .properties
+        .timestamp()
+        .and_then(|millis| Utc.timestamp_millis_opt(millis as _).latest())
+        .unwrap_or_else(Utc::now);
+
+    let routing = msg.routing_key.to_string();
+    let exchange = msg.exchange.to_string();
+    let keys = Keys {
+        routing_key: config.routing_key.as_str(),
+        exchange_key: config.exchange_key.as_str(),
+        offset_key: config.offset_key.as_str(),
+        routing: &routing,
+        exchange: &exchange,
+        delivery_tag: msg.delivery_tag as i64,
+    };
+
+    let stream = stream! {
+        while let Some(result) = stream.next().await {
+            match result {
+                Ok((events, byte_size)) => {
+                    emit!(AmqpBytesReceived {
+                        byte_size,
+                        protocol: "amqp_0_9_1",
+                    });
+
+                    emit!(AmqpEventsReceived {
+                        byte_size: byte_size,
+                        count: events.len(),
+                    });
+
+                    for mut event in events {
+                        populate_event(&mut event,
+                                       timestamp,
+                                       &keys,
+                                       log_namespace);
+
+                        yield event;
+                    }
+                }
+                Err(error) => {
+                    use codecs::StreamDecodingError as _;
```

Comment:
:thinking: is this necessary?

Reply:
Yeah, it's needed for the `can_continue` call.